File: java/open_redirect/CheckoutFlow.java
import jakarta.servlet.http.*;
import java.io.*;

public class CheckoutFlow extends HttpServlet {

    // VULN 1: Unvalidated `next` parameter used as redirect target after login
    public void handleLogin(HttpServletRequest req, HttpServletResponse resp) throws IOException {
        String username = req.getParameter("username");
        String password = req.getParameter("password");
        String next = req.getParameter("next");
        if (authenticateUser(username, password)) {
            resp.sendRedirect(next);
        } else {
            resp.sendRedirect("/login?error=invalid_credentials");
        }
    }

    // VULN 2: Unvalidated `return_url` parameter used as redirect after checkout completion
    public void handleCheckoutComplete(HttpServletRequest req, HttpServletResponse resp) throws IOException {
        String orderId = req.getParameter("order_id");
        String returnUrl = req.getParameter("return_url");
        processOrder(orderId);
        resp.sendRedirect(returnUrl);
    }

    // VULN 3: OAuth state parameter used directly as redirect target after OAuth callback
    public void handleOAuthCallback(HttpServletRequest req, HttpServletResponse resp) throws IOException {
        String code  = req.getParameter("code");
        String state = req.getParameter("state");
        exchangeOAuthCode(code);
        resp.sendRedirect(state);
    }

    private boolean authenticateUser(String u, String p) { return true; }
    private void processOrder(String orderId) {}
    private void exchangeOAuthCode(String code) {}
}

File: kotlin/open_redirect/CheckoutFlow.kt
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController

@RestController
class CheckoutFlow {

    // VULN 1: response.sendRedirect with unvalidated 'next' param after login
    @PostMapping("/auth/login")
    fun login(
        @RequestParam username: String,
        @RequestParam password: String,
        @RequestParam(defaultValue = "/shop") next: String,
        request: HttpServletRequest,
        response: HttpServletResponse
    ) {
        val session = request.getSession(true)
        session.setAttribute("username", username)
        response.sendRedirect(next)
    }

    // VULN 2: response.sendRedirect with unvalidated returnUrl after checkout completion
    @PostMapping("/checkout/complete")
    fun completeCheckout(
        @RequestParam orderId: String,
        @RequestParam(defaultValue = "/shop") returnUrl: String,
        request: HttpServletRequest,
        response: HttpServletResponse
    ) {
        val session = request.getSession(false)
        session?.removeAttribute("cart")
        response.sendRedirect(returnUrl)
    }

    // VULN 3: OAuth state parameter used directly as redirect target
    @GetMapping("/auth/oauth/callback")
    fun oauthCallback(
        @RequestParam code: String,
        @RequestParam state: String,
        response: HttpServletResponse
    ) {
        response.sendRedirect(state)
    }
}
